File: src/main/java/com/qtech/bubbles/registry/DeferredRegister.java
package com.qtech.bubbles.registry;

import com.qtech.bubbles.common.IRegistryEntry;
import com.qtech.bubbles.common.ResourceEntry;
import com.qtech.bubbles.common.mod.ModInstance;
import com.qtech.bubbles.event._common.SubscribeEvent;
import com.qtech.bubbles.event.bus.LocalAddonEventBus;
import com.qtech.bubbles.event.registry.RegistryEvent;
import com.qtech.bubbles.registry.object.RegistryObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.function.Supplier;

public class DeferredRegister<T extends IRegistryEntry> {
    private final String addonId;
    private final Registry<T> registry;
    private final ArrayList<HashMap.Entry<ResourceEntry, Supplier<T>>> objects = new ArrayList<>();

    protected DeferredRegister(String addonId, Registry<T> registry) {
        this.addonId = addonId;
        this.registry = registry;
    }

    public static <T extends RegistryEntry> DeferredRegister<T> create(String addonId, Registry<T> registry) {
        return new DeferredRegister<>(addonId, registry);
    }

    public <C extends T> RegistryObject<C> register(String key, Supplier<C> supplier) {
        ResourceEntry rl = new ResourceEntry(addonId, key);

//        if (!registry.getType().isAssignableFrom(supplier.get().getClass())) {
//            throw new IllegalArgumentException("Tried to register illegal type: " + supplier.get().getClass() + " expected assignable to " + registry.getType());
//        }

        objects.add(new HashMap.SimpleEntry<>(rl, supplier::get));

        return new RegistryObject<>(registry, supplier, rl);
    }

    public void register(LocalAddonEventBus<? extends ModInstance> eventBus) {
        eventBus.register(this);
    }

    @SubscribeEvent
    public void onRegister(RegistryEvent.Register<T> event) {
        if (!event.getRegistry().getType().equals(registry.getType())) {
            return;
        }

//        System.out.println("Deferred Register dump: " + event.getRegistry().getType());
        for (HashMap.Entry<ResourceEntry, Supplier<T>> entry : objects) {
            T object = entry.getValue().get();
            ResourceEntry rl = entry.getKey();

            if (!event.getRegistry().getType().isAssignableFrom(object.getClass())) {
                throw new IllegalArgumentException("Got invalid type in deferred register: " + object.getClass() + " expected assignable to " + event.getRegistry().getType());
            }

//            System.out.println("  (" + rl + ") -> " + object);

            event.getRegistry().register(rl, object);
            object.setRegistryName(rl);
        }
    }
}

File: src/main/kotlin/com/qtech/bubbles/registry/DeferredRegister.kt
package com.qtech.bubbles.registry

import com.qtech.bubbles.common.IRegistryEntry
import com.qtech.bubbles.common.ResourceLocation
import com.qtech.bubbles.common.addon.AddonInstance
import com.qtech.bubbles.event.SubscribeEvent
import com.qtech.bubbles.event.bus.LocalAddonEventBus
import com.qtech.bubbles.event.registry.RegistryEvent
import com.qtech.bubbles.registry.`object`.RegistryObject
import java.util.AbstractMap.SimpleEntry
import java.util.function.Supplier

class DeferredRegister<T : IRegistryEntry> private constructor(private val addonId: String?, private val registry: Registry<T>) {
    private val objects = ArrayList<Map.Entry<ResourceLocation, Supplier<T>>>()
    fun <C : T> register(key: String, supplier: Supplier<C>): RegistryObject<C> {
        val rl = ResourceLocation(addonId, key)

//        if (!registry.getType().isAssignableFrom(supplier.get().getClass())) {
//            throw new IllegalArgumentException("Tried to register illegal type: " + supplier.get().getClass() + " expected assignable to " + registry.getType());
//        }
        objects.add(SimpleEntry(rl, Supplier { supplier.get() }))
        return RegistryObject(registry, supplier, rl)
    }

    fun register(eventBus: LocalAddonEventBus<out AddonInstance>) {
        eventBus.register(this)
        eventBus.addListener(::onRegister)
    }

    @SubscribeEvent
    @JvmName("onRegister")
    fun onRegister(event: RegistryEvent.Register<T>) {
        if (event.registry.type != registry.type) {
            return
        }

        println("Deferred Register dump: " + event.registry.type);
        for ((rl, value) in objects) {
            val `object` = value.get()
            require(event.registry.type.isAssignableFrom(`object`.javaClass)) { "Got invalid type in deferred register: " + `object`.javaClass + " expected assignable to " + event.registry.type }

            println("  ($rl) -> $`object`");
            event.registry.register(rl, `object`)
            `object`.registryName = rl
        }
    }

    companion object {
        fun <T : IRegistryEntry> create(addonId: String?, registry: Registry<T>): DeferredRegister<T> {
            return DeferredRegister(addonId, registry)
        }
    }
}